components/Buttons/src/MDCButton.m

```diff
@@ -667,10 +667,16 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
 #pragma mark - Border Color
 
 - (UIColor *)borderColorForState:(UIControlState)state {
+  if ((state & UIControlStateHighlighted) == UIControlStateHighlighted) {
+    state = state & ~UIControlStateDisabled;
+  }
   return _borderColors[@(state)] ?: _borderColors[@(UIControlStateNormal)];
 }
 
 - (void)setBorderColor:(UIColor *)borderColor forState:(UIControlState)state {
+  if ((state & UIControlStateHighlighted) == UIControlStateHighlighted) {
```

Comment:
Looks like this is still missing some logic.  Please see #5919 for the logic required to match behavior.

```
// Only update the backing dictionary if:
// 1. The `state` argument is the same as the "storage" state, OR
// 2. There is already a value in the "storage" state.
```

Reply:
I am curious about the logic here:

> 
```
  // Only update the backing dictionary if:
  // 1. The `state` argument is the same as the "storage" state, OR
  // 2. There is already a value in the "storage" state.
  if (storageState == state || _backgroundColors[@(storageState)] != nil) {
    _backgroundColors[@(storageState)] = backgroundColor;
    [self updateAlphaAndBackgroundColorAnimated:NO];
  }
```
In this case, the `_backgroundColors` is not set when `(storageState != state && _backgroundColors[@(storageState)] == nil`, which only happens when state = highlight & disable. If we perform `state = state & ~UIControlStateDisabled;` highlight & disable also never gets set, which should behave the same unless _backgroundColor[highlight & disable] is accidentally set by some other method.

Did I understand it correctly?

Reply:
Not quite. The logic (simplified) is this:

1.  Attempt to set `.highlighted`
    1.  Always succeeds
1.  Attempt to set `.disabled`
    1.  Always succeeds
1.  Attempt to set `.highlighted | .disabled`
    1.  If `.highlighted` is set, override the value
    2.  If `.highlighted` is unset, do nothing
1.  Attempt to set `.selected | .highlighted | .disabled`
    1.  If `.selected | .highlighted` is set, override the value
    2.  If `.selected | .highlighted` is unset, do nothing

Reply:
I see. Thanks for the clarification:) The logic is interesting.

Addressed in 8e10afe46d3716fac6ca42aba4812b11d673bba3.